src/librustc/mir/interpret/allocation.rs

```diff
@@ -41,6 +45,481 @@ pub struct Allocation<Tag=(),Extra=()> {
     pub extra: Extra,
 }
 
+/// Alignment and bounds checks
+impl<'tcx, Tag, Extra> Allocation<Tag, Extra> {
+    /// Check that the pointer is aligned AND non-NULL. This supports ZSTs in two ways:
+    /// You can pass a scalar, and a `Pointer` does not have to actually still be allocated.
+    pub fn check_align(
+        &self,
+        ptr: Pointer<Tag>,
+        required_align: Align
+    ) -> EvalResult<'tcx> {
+        let size = Size::from_bytes(self.bytes.len() as u64);
+        if ptr.offset > size {
+            return err!(PointerOutOfBounds {
+                ptr: ptr.erase_tag(),
+                access: true,
+                allocation_size: size,
+            });
+        };
+        // Check alignment
+        if self.align.abi() < required_align.abi() {
+            return err!(AlignmentCheckFailed {
+                has: self.align,
+                required: required_align,
+            });
+        }
+        let offset = ptr.offset.bytes();
+        if offset % required_align.abi() == 0 {
+            Ok(())
+        } else {
+            let has = offset % required_align.abi();
+            err!(AlignmentCheckFailed {
+                has: Align::from_bytes(has, has).unwrap(),
+                required: required_align,
+            })
+        }
+    }
```

Comment:
`check_align` also still exists in memory.rs. Why have it twice?

Reply:
the `Memory` variant takes `Scalar`, while this one only takes `Pointer`, because the `Memory` version also works with dangling pointers and integer pointers

Reply:
Still, this is subtle code that we shouldn't duplicate.

Why have a `Pointer` version as well?

Reply:
`Allocation` needs such a method, but cannot possibly have access to the deallocated list. I tried finding some way to share code, but I'm unsure how to do this properly

Reply:
Why does `Allocation` need such a method? I do not even recall your first PR doing this duplication, but maybe I misremember?

We could say it is up to the caller to ensure that things are properly aligned, if they even care. Since the caller has to filter out ZSTs before doing `to_ptr()?`, they have to do an alignment check *anyway*.

Reply:
The duplication was there, too. My main worry is that if we do not do this, const processing might forget to do it $somewhere.

Reply:
Actual const *executions* will go through `Memory`. And when inspecting the result of a const execution after it is done, why should we care about alignment?

Reply:
Ok, ignoring this point, which I haven't put much thought into, we're still left with "removing `check_align` from `Allocation` will be very fragile because now `Memory` has to do all these checks in a billion places instead of being able to assume that `get_bytes_internal` and `get_bytes_mut` will do the checks for us (and forcing us to think about it by taking an `align` argument that needs to be fiddled through)

Reply:
Hm. How many exactly is a "billion"? Notice that any place that supports `Scalar` *already* has to do an alignment check in `Memory`, so I don't think it's that many...

Reply:
I made the changes in the last commit

Reply:
Meaning you force-pushed? :( Or which commit do you mean?

Reply:
Ah no not a force-push, I found it ("Move alignment checks out of `Allocation`"). But why did GitHub not send a notification?!?